src/rgw/rgw_http_client.h

```diff
@@ -102,6 +102,52 @@ class RGWHTTPClient
   int get_req_retcode();
 };
 
+
+class RGWHTTPHeadersCollector : public RGWHTTPClient {
+public:
+  /* Case insensitive comparator for containers carrying HTTP headers. */
+  struct CILess : public std::binary_function<std::string, std::string, bool> {
+    bool operator()(const std::string& lhs,
+                    const std::string& rhs) const {
+      return ::strcasecmp(lhs.c_str(), rhs.c_str()) < 0 ;
+    }
+  };
+
+  typedef std::string header_name_t;
+  typedef std::string header_value_t;
+  typedef std::set<header_name_t, CILess> header_spec_t;
+
+  RGWHTTPHeadersCollector(CephContext * const cct,
+                          const header_spec_t relevant_headers)
+    : RGWHTTPClient(cct),
+      relevant_headers(relevant_headers) {
+  }
+
+  int receive_header(void *ptr, size_t len) override;
```

Comment:
Those methods are called by callbacks (like `[simple_]receive_http_header`) we passed to libcurl.